src/libraries/Microsoft.PowerFx.Core/Public/Config/DeferredSymbolTable.cs

```diff
@@ -0,0 +1,165 @@
+using Microsoft.PowerFx.Core.Functions;
+using Microsoft.PowerFx.Core.Types.Enums;
+using Microsoft.PowerFx.Core.Utils;
+using Microsoft.PowerFx.Types;
+
+namespace Microsoft.PowerFx
+{
+    /// <summary>
+    /// Provides symbol table with deferred semantics. 
+    /// This is semantically readonly - there is a fixes set of known symbols, just the type information is populated lazily.
+    /// Good when we can quickly load the names (to populate the intellisense completion list), but then load remaining symbol details on-demand. 
+    /// </summary>
+    [DebuggerDisplay("{DebugName}")]
+    internal class DeferredSymbolTable : ReadOnlySymbolTable, IGlobalSymbolNameResolver, INameResolver
+    {
+        // All possible tables we could add. 
+        private readonly DisplayNameProvider _displayNameLookup;
+
+        // (Logical,Display) --> return type. 
+        // Call back to host to a) lazily get the type, b) notify the host this has been loaded. 
+        // Assume it can be invoked multiple times; host must protect
+        private readonly Func<string, string, FormulaType> _fetchTypeInfo;
+
+        // Full universe of possible symbols
+        // All other symbols are missing. 
+        public DeferredSymbolTable(
+            DisplayNameProvider map, 
+            Func<string, string, FormulaType> fetchTypeInfo)
+        {
+            _displayNameLookup = map ?? throw new ArgumentNullException(nameof(map));
+            _fetchTypeInfo = fetchTypeInfo ?? throw new ArgumentNullException(nameof(fetchTypeInfo));
+        }
+
+        // SymbolTable is conceptually constant. 
+        private readonly VersionHash _constant = VersionHash.New();
+
+        internal override VersionHash VersionHash => _constant;
+
+        // Must be thread safe!!!        
+        // We can have multiple threads reading; which means they may be populating the cache. 
+        private readonly object _lock = new object();
+
+        private readonly Dictionary<string, NameLookupInfo> _variables = new Dictionary<string, NameLookupInfo>();
+
+        private readonly SlotMap<NameLookupInfo?> _slots = new SlotMap<NameLookupInfo?>();
+
+        // Override lookup.
+        bool INameResolver.Lookup(DName name, out NameLookupInfo nameInfo, NameLookupPreferences preferences)
+        {
+            if (_displayNameLookup.TryGetLogicalOrDisplayName(name, out var logical, out var display))
+            {
+                lock (_lock)
+                {
+                    // Previously cached 
+                    if (_variables.TryGetValue(logical.Value, out nameInfo))
+                    {
+                        return true;
+                    }
+                }
+
+                // Callback to host - do not hold lock while invoking this. 
+                // Host may do arbitrary operations, including network calls to fetch metadata.
+                var type = _fetchTypeInfo(logical.Value, display.Value);
+
+                lock (_lock)
```

Comment:
>lock (_lock) [](http://example.com/codeflow?start=16&length=12)

Why not put lock inside method `AddUnderLock()`?

Reply:
It's a style thing - I wanted all the lock/unlock to be at the same level so that it was clear when it was being taken and released. It's especially important that we don't hold the lock while invoking callbacks (and virtuals implemented by the host are also callbacks).